### DISCUSSION TITLE
Websocket stream spot, chronological order of data?

### DISCUSSION THREAD

**Discussion by CharlyEmpereurmot** [oliver-zehentleitner/unicorn-binance-websocket-api#270] [General]
Hi! So I have been using this package for a while, thanks again for this Oliver :)

I'm applying some machine learning models to predict the next price action for a given period, just UP or DOWN. My models are calibrated according to historical data, then I stream (spot) data from Binance and make predictions, act accordingly. The predictions yield a continuous value within `[0, 1]` and I have a hard threshold that says for example `if prediction > 0.55, then next_price_action = UP`.

What is puzzling me is that I see some discrepencies in the predictions my models are making, when it's using historical data VS the live streamed data. Sometimes I will get `prediction = 0.56` using historical data and from the live streamed data I will get `prediction = 0.54`. After investigating quite a lot, I don't think the problem comes from my code, but rather from the underlying data the models are fed. I'm NOT using a callback, but rather `stream_payload = binance_websocket_api_manager.pop_stream_data_from_stream_buffer()`. And I trigger a new predicton as soon as I see a new timestamp for a candle opening, on a given pair I want to trade.

So I'm thinking now, the problem is that Binance does NOT send the candle information in the exact chronological order and/or these data do NOT come to my server in the correct chronological order, for whatever else reason.

Have you guys experienced similar behavior of the websocket streamed data? I have seen around that this is to be expected, but I did not think this would be so frequent. I had understood this could happen only "under Binance servers heavy load periods". What are you guys doing to handle this kind of problem?

Thank you very much for any advices you may provide!

**Comment by oliver-zehentleitner** [comment]
Hi Charly!

if you use historical data, you have only full candle sticks. (the closing one of each interval you used)

if you subscribe to websocket you get an opening candle with 0 vol i think and then you receive updates if there are trades. so this is different from trainging data.

you need to record this data to a database and train your network to it if it should be able to use data between two closing candles.

the last kline should have is_closed = True

Between 2 closing candles is minimum 1 minute, for me i think this never happend to be in the wrong order....

did you try to log your received candles and controll it?

**Reply by CharlyEmpereurmot** [reply]
Hey :) Yes that's what I'm doing, recording to a DB and then just as soon as I see a new candle opening, I will trigger a prediction to be made from the past data, which are indeed identical as the historical data (this I have verified, but only in between events of "candle opening").

I am just checking for an un-seen open timestamp though, to declare a new "candle opening" event. And I'm NOT checking for is_closed = True on the previous candle. I might want to do this, will try and let you know. Currently I'm not logging the received candles but I'm starting to check the receival order now.

Thanks, I will be back here for updates!

**Reply by CharlyEmpereurmot** [reply]
Hi Oliver! So now I think my problem simply boils down to some sort of shifted candles timestamp because of different timezones between my EU server and Binance ones. I will look into how to handle this properly. Especially in case the change of hour is not in sync in between the 2x timezones.

**Reply by MoeQatoum** [reply]
hi Charly, I've just seen this discussion and I want to point out that sometimes kline msg with is_closed = True is not sent from binance side, especially at times of high volatility. I wouldn't just rely on is_closed = True kline msg , you would be better off writing logic to handle this issue.

**Reply by CharlyEmpereurmot** [reply]
Hey! Ok, thanks for the additional observations.
Well in fact I tried something in this sense at first. Please see also the beginning of the (short) discussion.
Maybe I should have logic to keep track of the `Event time` provided in `payload['e']` and that would do the job. That data is the timestamp of the event, correct? How do you guys proceed to avoid issues at this level?
I would gladly take any other piece of advice you would input! Cheers

**Reply by oliver-zehentleitner** [reply]
Thanks for this hint @MoeQatoum 

To prove without `is_closed` its possible to do `stream_data['event_time'] >= stream_data[kline']['kline_close_time']`.

**Reply by MoeQatoum** [reply]
@oliver-zehentleitner @CharlyEmpereurmot 
yup this is how I'm doing it, if I receive `TradeMsg` with`msg.E >= df.iloc[-1].closetime`, current kline is closed and new kline is added to df, then process current `TradeMsg`

what I noticed also that sometimes `TradeMsg` for next kline is received before the `is_closed=True` `KlineMsg` is received, due to `KlineMsg` update speed as mentioned in [binance documentation](https://binance-docs.github.io/apidocs/spot/en/#kline-candlestick-streams) is sent every `2 sec` while `TradeMsg` update speed is [real time](https://binance-docs.github.io/apidocs/spot/en/#trade-streams). that's another reason why you shouldn't rely on `is_closed=True` `KlineMsg` for closing validation.

**Reply by CharlyEmpereurmot** [reply]
Thanks for the help :)

Now I do see indeed these events in my logs:
```
BTCUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371200003 exceeding the closing ts, closing candle and ignoring data
OMGUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371200197 exceeding the closing ts, closing candle and ignoring data
VETUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371200369 exceeding the closing ts, closing candle and ignoring data
VETUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371200370 exceeding the closing ts, closing candle and ignoring data
ALGOUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371200487 exceeding the closing ts, closing candle and ignoring data
RVNUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371201339 exceeding the closing ts, closing candle and ignoring data
BATUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371201367 exceeding the closing ts, closing candle and ignoring data
KAVAUSDT_8h ---- WARNING: Received data for candle open_ts 1652342400000 @ ts 1652371201370 exceeding the closing ts, closing candle and ignoring data
```

The 2nd timestamp is the `event_time`. Now I monitor for both `is_closed` and the `event_time` to which the candle data corresponds, when it exceeds the `ts_close` of the candle then I ignore this given data point for the candle and I declare it closed. Then I will trigger a prediction at the very next data point received for the newly opened candle, using the data I have stored in DB previously, and hopefully the data received for the previous candle(s) is indeed identical to historical data when queried slightly later than at the very moment the new candle opens.

It's a bit puzzling to me that for VET there were 2 data points at just 1 ms difference... But OK, I guess it's just Binance-internal-thingy-related. `event_time` is supposed to be the time the data point was generated/sent from their servers, correct?

I will monitor the behavior, thanks a lot for your help guys! Big time

**Reply by MoeQatoum** [reply]
Welc. Yeah I guess it's binance internal issue and nowadays is very nice time to stress test your code the market is very volatile.

Please let us know how is ai doing with guessing price direction. I always thought it's impossible for ai to predict price direction as it's completely random, so never tried it. Please let us know the results.

**Reply by oliver-zehentleitner** [reply]
A very good resource to this topic: https://github.com/PacktPublishing/Machine-Learning-for-Algorithmic-Trading-Second-Edition

**Reply by CharlyEmpereurmot** [reply]
It's not an easy task, but there definitely are ways. Even in randomness (or close to it, anyways) you can isolate some signals, but you need to find a way to select the strongest ones among the noise. Then forget about 70% accuracy/precision, this cannot happen (or very rare cases, not stable long-term). But you can already do stuff with > 53% for trading.

**Reply by oliver-zehentleitner** [reply]
I think the key is the training data :) all depends from good classifications

What good or bad signals are is the task of the learning algorithm.

**Reply by CharlyEmpereurmot** [reply]
Yes, and of the metrics used! What do you mean exactly about the training data? Something like using 3 classes (having a neutral one) instead of binary classifying? Maybe also using stratification for the cross-validation? I would love to hear a few more words about that :)
